lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -4245,6 +4245,96 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
+          BM->addInstTemplate(OpSignBitSet, {TestInst->getId()}, BB, Ty);
+      if (IsNegative)
+        return SignBitTest;
+      return BM->addInstTemplate(OpLogicalNot, {SignBitTest->getId()}, BB, Ty);
+    };
+    // Integer paramter of the intrinsic is combined from several bit masks
+    // referenced in FPClassTest enum from FloatingPointMode.h in LLVM.
+    // Since a single intrinsic can provide multiple tests - here we might end
+    // up adding several sequences of SPIR-V instructions
+    if (FPClass & fcSNan || FPClass & fcQNan) {
+      // There are no signaling and quite NaN instructions in SPIR-V, so map
+      // both of the tests on OpIsNan
+      ResultVec.emplace_back(BM->addInstTemplate(OpIsNan, Ops, BB, Ty));
+    }
+    if (FPClass & fcNegInf || FPClass & fcPosInf) {
+      auto *TestIsInf = BM->addInstTemplate(OpIsInf, Ops, BB, Ty);
+      if (FPClass & fcNegInf && FPClass & fcPosInf)
+        // Map on OpIsInf if we have both Inf test bits set
+        ResultVec.emplace_back(TestIsInf);
+      else
+        // Map on OpIsInf with following check for sign bit
+        ResultVec.emplace_back(GetNegPosInstTest(TestIsInf,
+                                                 FPClass & fcNegInf));
+    }
+    if (FPClass & fcNegNormal || FPClass & fcPosNormal) {
+      auto *TestIsNormal = BM->addInstTemplate(OpIsNormal, Ops, BB, Ty);
+      if (FPClass & fcNegNormal && FPClass & fcPosNormal)
+        // Map on OpIsNormal if we have both Normal test bits set
+        ResultVec.emplace_back(TestIsNormal);
+      else
+        // Map on OpIsNormal with following check for sign bit
+        ResultVec.emplace_back(GetNegPosInstTest(TestIsNormal,
+                                                 FPClass & fcNegNormal));
+    }
+    if (FPClass & fcNegSubnormal || FPClass & fcPosSubnormal) {
+      // If value is both not NaN and Normal, then it's subnormal
+      // TODO: it's quite silly check, can we figure out something better?
+      auto *TestIsNan = BM->addInstTemplate(OpIsNan, Ops, BB, Ty);
+      auto *TestIsNormal = BM->addInstTemplate(OpIsNormal, Ops, BB, Ty);
+      auto *TestIsNotSubnormal =
+          BM->addInstTemplate(OpLogicalOr, {TestIsNan->getId(),
+                                            TestIsNormal->getId()}, BB, Ty);
+      auto *TestIsSubnormal =
+          BM->addInstTemplate(OpLogicalNot, {TestIsNotSubnormal->getId()}, BB,
+                              Ty);
+      if (FPClass & fcNegSubnormal && FPClass & fcPosSubnormal)
+        ResultVec.emplace_back(TestIsSubnormal);
+      else
+        ResultVec.emplace_back(GetNegPosInstTest(TestIsSubnormal,
+                                                 FPClass & fcNegZero));
+    }
+    if (FPClass & fcNegZero || FPClass & fcPosZero) {
+      // Map on OpFOrdEqual compare to 0.0 with following check for sign bit
+      Constant *Zero = ConstantFP::getZero(II->getArgOperand(0)->getType());
+      auto *TestIsZero = BM->addCmpInst(OpFOrdEqual, Ty, InputFloat,
```

Comment:
Adding OpFOrdEqual is also a bit questionable, since it's instruction from SPIR-V 1.6 and we officially support only SPIR-V 1.4 in the translator.